### ISSUE TITLE
to many open files on project with a lot of files

### ISSUE THREAD

**Issue by TomDePauw** [nebula-plugins/gradle-lint-plugin#91]
Hello, i have an issue when running applying the gradle lint to a project with a lot of files (around 24.057).
Please see https://github.com/gradle/gradle/issues/1374 . I am not sure where the problem is.
A lot of files in our project are not java files (around 18 000 are not java related) .

**Comment by ketan** [comment]
Bump... I'm seeing this issue on my current project with several files

**Comment by ayavorsk2000-gmail** [comment]
Same issue here. Trying to run on a very large java code base (3M lines of code) and lint process fails with "Two many files open" error. Not only that, but after encountering the error I have to reboot my Mac, as it severely depletes machine resources such that even Chrome browser cannot start.

It would be great if you could prioritize the fix for this issue. The tool is very promising and I am trying to see if it can help us fix a 250+ dependency set.

**Comment by DanielThomas** [comment]
Does someone have a stacktrace that demonstrates the issue so I can get an idea of where we're seeing the file handle leak?

**Comment by ketan** [comment]
Here you go —

```
Error from [com.netflix.nebula.lint.rule.GradleLintRule$1] processing source file [null]
:lintGradle FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':lintGradle'.
> java.nio.file.FileSystemException: /Users/ketanpadegaonkar/projects/gocd/gocd/server/target/classes/test/com/thoughtworks/go/server/service/plugins/processor/authentication: Too many open files in system

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':lintGradle'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:69)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.PostExecutionAnalysisTaskExecuter.execute(PostExecutionAnalysisTaskExecuter.java:35)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:64)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:53)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:203)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:185)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.processTask(AbstractTaskPlanExecutor.java:74)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.run(AbstractTaskPlanExecutor.java:55)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
        at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
Caused by: org.gradle.api.UncheckedIOException: java.nio.file.FileSystemException: /Users/ketanpadegaonkar/projects/gocd/gocd/server/target/classes/test/com/thoughtworks/go/server/service/plugins/processor/authentication: Too many open files in system
        at org.gradle.internal.UncheckedException.throwAsUncheckedException(UncheckedException.java:43)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:78)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.doExecute(DefaultTaskClassInfoStore.java:133)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:126)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:115)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:623)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:606)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:80)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:61)
        ... 14 more
Caused by: java.nio.file.FileSystemException: /Users/ketanpadegaonkar/projects/gocd/gocd/server/target/classes/test/com/thoughtworks/go/server/service/plugins/processor/authentication: Too many open files in system
        at java_nio_file_Files$walkFileTree.call(Unknown Source)
        at com.netflix.nebula.lint.rule.dependency.DependencyService.memoizedMethodPriv$classReferencesString(DependencyService.groovy:156)
        at com.netflix.nebula.lint.rule.dependency.DependencyService$_closure2.doCall(DependencyService.groovy)
        at com.netflix.nebula.lint.rule.dependency.DependencyService.classReferences(DependencyService.groovy)
        at com.netflix.nebula.lint.rule.dependency.DependencyService.memoizedMethodPriv$usedDependenciesString(DependencyService.groovy:268)
        at com.netflix.nebula.lint.rule.dependency.DependencyService$_closure6.doCall(DependencyService.groovy)
        at com.netflix.nebula.lint.rule.dependency.DependencyService.usedDependencies(DependencyService.groovy)
        at com.netflix.nebula.lint.rule.dependency.DependencyService$usedDependencies$13.call(Unknown Source)
        at com.netflix.nebula.lint.rule.dependency.UnusedDependencyRule$_visitGradleDependency_closure1.doCall(UnusedDependencyRule.groovy:58)
        at com.netflix.nebula.lint.rule.dependency.UnusedDependencyRule.visitGradleDependency(UnusedDependencyRule.groovy:57)
        at com.netflix.nebula.lint.rule.GradleLintRule.this$dist$invoke$2(GradleLintRule.groovy)
        at com.netflix.nebula.lint.rule.GradleLintRule$1.methodMissing(GradleLintRule.groovy)
        at com.netflix.nebula.lint.rule.GradleLintRule$1.this$dist$invoke$3(GradleLintRule.groovy)
        at com.netflix.nebula.lint.rule.GradleLintRule$1$2.methodMissing(GradleLintRule.groovy)
        at com.netflix.nebula.lint.rule.GradleLintRule$1$2.visitMethodCallInDependencies(GradleLintRule.groovy:402)
        at com.netflix.nebula.lint.rule.GradleLintRule$1$2.visitMethodCallExpression(GradleLintRule.groovy:224)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor$_visitMethodCallExpression_closure31.doCall(CompositeGroovyAstVisitor.groovy:284)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.visitMethodCallExpression(CompositeGroovyAstVisitor.groovy:284)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.super$3$visitExpressionStatement(CompositeGroovyAstVisitor.groovy)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.visitExpressionStatement(CompositeGroovyAstVisitor.groovy:225)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.super$3$visitBlockStatement(CompositeGroovyAstVisitor.groovy)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.visitBlockStatement(CompositeGroovyAstVisitor.groovy:189)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.super$2$visitClosureExpression(CompositeGroovyAstVisitor.groovy)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.visitClosureExpression(CompositeGroovyAstVisitor.groovy:347)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.super$2$visitListOfExpressions(CompositeGroovyAstVisitor.groovy)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.visitListOfExpressions(CompositeGroovyAstVisitor.groovy:473)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.super$2$visitTupleExpression(CompositeGroovyAstVisitor.groovy)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.visitTupleExpression(CompositeGroovyAstVisitor.groovy:353)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.super$2$visitArgumentlistExpression(CompositeGroovyAstVisitor.groovy)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.visitArgumentlistExpression(CompositeGroovyAstVisitor.groovy:479)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.super$2$visitMethodCallExpression(CompositeGroovyAstVisitor.groovy)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.visitMethodCallExpression(CompositeGroovyAstVisitor.groovy:286)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.super$3$visitExpressionStatement(CompositeGroovyAstVisitor.groovy)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.visitExpressionStatement(CompositeGroovyAstVisitor.groovy:225)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.super$3$visitBlockStatement(CompositeGroovyAstVisitor.groovy)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.visitBlockStatement(CompositeGroovyAstVisitor.groovy:189)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.super$3$visitClassCodeContainer(CompositeGroovyAstVisitor.groovy)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.visitClassCodeContainer(CompositeGroovyAstVisitor.groovy:136)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.super$3$visitConstructorOrMethod(CompositeGroovyAstVisitor.groovy)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.visitConstructorOrMethod(CompositeGroovyAstVisitor.groovy:147)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.super$3$visitMethod(CompositeGroovyAstVisitor.groovy)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.visitMethod(CompositeGroovyAstVisitor.groovy:159)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.super$3$visitClass(CompositeGroovyAstVisitor.groovy)
        at com.netflix.nebula.lint.rule.CompositeGroovyAstVisitor.visitClass(CompositeGroovyAstVisitor.groovy:105)
        at com.netflix.nebula.lint.org.codenarc.rule.AbstractAstVisitorRule.applyTo(AbstractAstVisitorRule.java:97)
        at com.netflix.nebula.lint.org.codenarc.rule.AbstractRule.applyTo(AbstractRule.java:142)
        at com.netflix.nebula.lint.org.codenarc.rule.Rule$applyTo$0.call(Unknown Source)
        at com.netflix.nebula.lint.rule.GradleLintRule.applyTo(GradleLintRule.groovy:158)
        at com.netflix.nebula.lint.org.codenarc.rule.Rule$applyTo$0.call(Unknown Source)
        at com.netflix.nebula.lint.org.codenarc.rule.Rule$applyTo$0.call(Unknown Source)
        at com.netflix.nebula.lint.org.codenarc.analyzer.AbstractSourceAnalyzer.collectViolations(AbstractSourceAnalyzer.groovy:40)
        at com.netflix.nebula.lint.plugin.LintService$ReportableAnalyzer.analyze(LintService.groovy:48)
        at com.netflix.nebula.lint.plugin.LintService$ReportableAnalyzer$analyze.call(Unknown Source)
        at com.netflix.nebula.lint.plugin.LintService$_lint_closure4.doCall(LintService.groovy:108)
        at com.netflix.nebula.lint.plugin.LintService.lint(LintService.groovy:105)
        at com.netflix.nebula.lint.plugin.LintService$lint.call(Unknown Source)
        at com.netflix.nebula.lint.plugin.LintGradleTask.lint(LintGradleTask.groovy:34)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:75)
        ... 21 more
```

**Comment by eliezio** [comment]
I was able to reproduce this problem under the following conditions:
- I have a project with 446 class files
- on build.gradle I set: gradleLint.rules = ['unused-dependency']

If I run gradle on daemon mode, the daemon keeps 834 class files opened! (because some of them are opened twice!)

I also confirmed that this leakage does NOT occur if I only use the 'duplicate-dependency-class' rule.

To avoid this leakage problem I'm running gradle with '--no-daemon'.

**Comment by ayavorsk2000-gmail** [comment]
Unfortunately, --no-daemon option does not solve it in my case. I still get the too many open files error as shown below.

:autoLintGradle
Error from [com.netflix.nebula.lint.rule.GradleLintRule$1] processing source file [null]
:autoLintGradle FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':autoLintGradle'.
> java.nio.file.FileSystemException: /Users/ayavorskiy/git/gitolith/domain/build/classes/main/domain/adminweb/printconfig/salessample: Too many open files in system

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

**Comment by DanielThomas** [comment]
It's definitely the code indicated in the stacktrace provided by @ketan. `ClassReader` doesn't close the `InputStream` (the false parameter below):

    public ClassReader(final InputStream is) throws IOException {
        this(readClass(is, false));
    }

Expect a patch release later today.

**Comment by DanielThomas** [comment]
Fixed by v7.2.3.
